### DISCUSSION TITLE
Verifying server certificate by OpenSSL

### DISCUSSION THREAD

**Discussion by assen-totin** [strongswan/strongswan#1786] [Q&A]
Hello! 

We have a somewhat interesting case revolving around being friendly to the mass user.

We have a relatively simple setup:

- StrongSwan server 5.9 (RHEL) which identifies itself to its clients with a certificate, signed by a public authority. StrongSwan is also configured to push the CA's intermediate certificate, which works fine.
- Linux road warrior clients use NetworkManager + StrongSwan (via charon-nm) to connect on-the-go, which need to verify the server's certificate.

If the place the root of the CA as a single PEM file into StrongSwan's ca-specific directory (like /etc/strongswan/ipsec.d/cacerts on RHEL/Fedora and the likes), it all works fine. However, this requires the client to manually save the CA root into a specific location.

We know that the very same CA root is available in the OS-provided certificate store (under /etc/pki in our case), so we would, naturally, be happy if StrongSwan could use this store to automatically verify the server's certificate:

- The OS provides exports from its credentials store in various formats; there is an all-in-one PEM bundle export - but when we symlink it in the StrongSwan's CA directory (or when we configure charon-nm.ca_dir to point to its directory), server certificate verification fails; we spotted somewhere a remark that StrongSwan only reads the first entry from such a file, so this could be the reason.
- If we use the OpenSSL CLI to verify the server certificate together with the CA intermediate, it works fine (i.e. smth. like "openssl --verify --untrusted intermediate.pem server.pem").

So, the question is, do we have a way to ask StrongSwan to offload the server verification to OpenSSL, meaning ask OpenSSL to do it itself (using its CA store), instead of merely using OpenSSL as crypto provider and verifying the certificate on its own? If the latter is possible, will OCSP still function the usual (to us) way, i.e. check the revocation status online (on the CA web site) on-the-fly before accepting it?

To simplify the test, we ran the client (Fedora 38 - latest at the time of writing) with SELinux turned off.

Any hints/ideas are welcome. If we missed something while trying to link the OS-provided CA root export into StrongSwan's directory, advice will be appreciated.

Assen Totin

**Comment by tobiasbrunner** [comment]
> we spotted somewhere a remark that StrongSwan only reads the first entry from such a file, so this could be the reason.

Yes, see [the FAQs](https://docs.strongswan.org/docs/5.9/support/faq.html#_x_509_certificate_chain_files) (applies to all such bundles of PEM-encoded certificates).

> So, the question is, do we have a way to ask StrongSwan to offload the server verification to OpenSSL, meaning ask OpenSSL to do it itself (using its CA store), instead of merely using OpenSSL as crypto provider and verifying the certificate on its own?

No, that's not possible.

Doesn't `/etc/pki/ca-trust/extracted/pem/directory-hash/` contain the extracted certificates (I just spun up a Fedora container where that was the case)?

Also, maybe you could use the PKCS#11 trust module (`p11-kit-trust.so`) [provided by p11-kit on Fedora](https://fedoraproject.org/wiki/Features/SharedSystemCertificates).

**Comment by assen-totin** [comment]
Hello, Tobias,

Thank you for replying so fast.

> Doesn't `/etc/pki/ca-trust/extracted/pem/directory-hash/` contain the extracted certificates (I just spun up a Fedora container where that was the case)?

Yes, correct - we've missed this. I guess this is as close as we're going to get on RHEL (since their StrongSwan maintainer somehow distrusts trusted roots and some time ago refused to point --with-nm-ca-dir to a place under /etc/pki).

For posterity, it suffices to create `/etc/strongswan/strongswan.d/charon-nm.conf` with the following content:

```
charon-nm {
	ca_dir = /etc/pki/ca-trust/extracted/pem/directory-hash
}
```
